### ISSUE TITLE
mongo - $push not working with $ne selector

### ISSUE THREAD

**Issue by mcbain** [meteor/meteor#5294]
On the server side minimongo will be used to validate Mongo.Collection.update arguments?!.

The following upsert fails with `MinimongoError: Cannot apply $push modifier to non-array`.
Minimongo IMHO infers `abc` to be a String, but $ne is meant as array-contains selector in that case.

``` javascript
upsert({
    'abc': {'$ne': 444}
  },
  {
    '$push': {
      'abc': 123
    }
  })
```

**Comment by tmeasday** [comment]
Hi @mcbain can you post a reproduction of this? I would be surprised if this was the case.

**Comment by mcbain** [comment]
@tmeasday Made a repro meteor package with a tinytest to show the bug https://github.com/mcbain/meteor-mongo-bug1

Doing the upsert in a mongo shell works.

**Comment by tmeasday** [comment]
Hi @mcbain, thanks!

So you can see why Minimongo code is used for a server upsert here:
https://github.com/meteor/meteor/blob/devel/packages/mongo/mongo_driver.js#L525-L536

Explanation of why here: https://github.com/meteor/meteor/issues/2278#issuecomment-64252706

**Comment by tmeasday** [comment]
(I guess this at the end of the day is a limitation of Minimongo, that is triggered by server upserts)

**Comment by mcbain** [comment]
@tmeasday thanks for investigation. Since current mongo version is 3.2, is there a deprecation path for the 2.4 branch?

**Comment by tmeasday** [comment]
Good question. I think there is some work ongoing about getting us working properly with 3.0 but I'm not sure about deprecating our support for 2.4.

**Comment by mitar** [comment]
I can reproduce this even in the recent Meteor version. We should probably check if MongoDB is 2.4 or earlier, and use simulation only then, otherwise pass upsert directly to MongoDB.

**Comment by sebakerckhof** [comment]
Okay, this should be fixed in next 1.6 build now that https://github.com/meteor/meteor/issues/8815 is merged.

**Comment by hwillson** [comment]
Great, closing for now - thanks for your work on this @sebakerckhof!
